src/storage/src/hummock/shared_buffer/shared_buffer_batch.rs

```diff
@@ -48,20 +48,35 @@ pub(crate) struct SharedBufferBatchInner {
 impl SharedBufferBatchInner {
     fn new(
         payload: Vec<SharedBufferItem>,
-        delete_range_tombstones: Vec<DeleteRangeTombstone>,
+        mut range_tombstone_list: Vec<DeleteRangeTombstone>,
         size: usize,
         _tracker: Option<MemoryTracker>,
     ) -> Self {
         let mut largest_user_key = vec![];
-        for tombstone in &delete_range_tombstones {
-            // Although `end_user_key` of tombstone is exclusive, we still use it as a boundary of
-            // `SharedBufferBatch` because it just expands an useless query and does not affect
-            // correctness.
-            if largest_user_key.lt(&tombstone.end_user_key) {
-                largest_user_key.clear();
-                largest_user_key.extend_from_slice(&tombstone.end_user_key);
+        if !range_tombstone_list.is_empty() {
+            range_tombstone_list.sort();
+            let mut range_tombstones: Vec<DeleteRangeTombstone> = vec![];
+            for tombstone in range_tombstone_list {
+                // Although `end_user_key` of tombstone is exclusive, we still use it as a boundary
+                // of `SharedBufferBatch` because it just expands an useless query
+                // and does not affect correctness.
+                if largest_user_key.lt(&tombstone.end_user_key) {
+                    largest_user_key.clear();
+                    largest_user_key.extend_from_slice(&tombstone.end_user_key);
+                }
+                if let Some(last) = range_tombstones.last_mut() {
+                    if last.end_user_key.gt(&tombstone.start_user_key) {
+                        if last.end_user_key.lt(&tombstone.end_user_key) {
+                            last.end_user_key = tombstone.end_user_key;
+                        }
```

Comment:
again, why can we omit a tombstone with unknown sequence?

Reply:
No. The sequence must be right. Here we assign sequence for new tombstone by the origin tombstone.